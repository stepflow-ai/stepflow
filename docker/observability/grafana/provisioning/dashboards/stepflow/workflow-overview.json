{
  "annotations": {
    "list": [
      {
        "builtIn": 1,
        "datasource": {
          "type": "grafana",
          "uid": "-- Grafana --"
        },
        "enable": true,
        "hide": true,
        "iconColor": "rgba(0, 211, 255, 1)",
        "name": "Annotations & Alerts",
        "type": "dashboard"
      }
    ]
  },
  "editable": true,
  "fiscalYearStartMonth": 0,
  "graphTooltip": 1,
  "id": null,
  "links": [],
  "panels": [
    {
      "datasource": {
        "type": "prometheus",
        "uid": "prometheus"
      },
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "thresholds"
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              }
            ]
          },
          "unit": "short"
        },
        "overrides": []
      },
      "gridPos": {
        "h": 8,
        "w": 12,
        "x": 0,
        "y": 0
      },
      "id": 1,
      "options": {
        "colorMode": "value",
        "graphMode": "area",
        "justifyMode": "auto",
        "orientation": "auto",
        "reduceOptions": {
          "values": false,
          "calcs": [
            "lastNotNull"
          ],
          "fields": ""
        },
        "textMode": "auto"
      },
      "pluginVersion": "10.3.3",
      "targets": [
        {
          "datasource": {
            "type": "prometheus",
            "uid": "prometheus"
          },
          "expr": "sum(rate(stepflow_workflow_executions_total[5m]))",
          "refId": "A"
        }
      ],
      "title": "Workflow Execution Rate (per sec)",
      "type": "stat"
    },
    {
      "datasource": {
        "type": "prometheus",
        "uid": "prometheus"
      },
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "axisBorderShow": false,
            "axisCenteredZero": false,
            "axisColorMode": "text",
            "axisLabel": "",
            "axisPlacement": "auto",
            "barAlignment": 0,
            "drawStyle": "line",
            "fillOpacity": 10,
            "gradientMode": "none",
            "hideFrom": {
              "tooltip": false,
              "viz": false,
              "legend": false
            },
            "insertNulls": false,
            "lineInterpolation": "linear",
            "lineWidth": 1,
            "pointSize": 5,
            "scaleDistribution": {
              "type": "linear"
            },
            "showPoints": "never",
            "spanNulls": false,
            "stacking": {
              "group": "A",
              "mode": "none"
            },
            "thresholdsStyle": {
              "mode": "off"
            }
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              }
            ]
          },
          "unit": "s"
        },
        "overrides": []
      },
      "gridPos": {
        "h": 8,
        "w": 12,
        "x": 12,
        "y": 0
      },
      "id": 2,
      "options": {
        "legend": {
          "calcs": [],
          "displayMode": "list",
          "placement": "bottom",
          "showLegend": true
        },
        "tooltip": {
          "mode": "multi",
          "sort": "none"
        }
      },
      "pluginVersion": "10.3.3",
      "targets": [
        {
          "datasource": {
            "type": "prometheus",
            "uid": "prometheus"
          },
          "expr": "histogram_quantile(0.95, sum(rate(stepflow_workflow_duration_seconds_bucket[5m])) by (le))",
          "legendFormat": "p95",
          "refId": "A"
        },
        {
          "datasource": {
            "type": "prometheus",
            "uid": "prometheus"
          },
          "expr": "histogram_quantile(0.50, sum(rate(stepflow_workflow_duration_seconds_bucket[5m])) by (le))",
          "legendFormat": "p50",
          "refId": "B"
        }
      ],
      "title": "Workflow Duration (p50, p95)",
      "type": "timeseries"
    },
    {
      "datasource": {
        "type": "datasource",
        "uid": "-- Mixed --"
      },
      "gridPos": {
        "h": 4,
        "w": 24,
        "x": 0,
        "y": 8
      },
      "id": 3,
      "options": {
        "mode": "markdown",
        "content": "## Workflow Traces\n\nView distributed traces for workflow executions:\n\n- [Browse traces in Jaeger UI](http://localhost:16686/search?service=stepflow-cli)\n- [Explore traces in Grafana](/explore?left=%7B%22datasource%22%3A%22jaeger%22%2C%22queries%22%3A%5B%7B%22refId%22%3A%22A%22%2C%22queryType%22%3A%22search%22%2C%22service%22%3A%22stepflow-cli%22%7D%5D%7D)\n- Click on **TraceID** links in the logs panel below to view individual traces\n\n**Tip**: Logs are automatically linked to traces via `trace_id` - click any trace ID in the logs to view the full trace."
      },
      "pluginVersion": "10.3.3",
      "targets": [],
      "title": "Trace Navigation",
      "type": "text"
    },
    {
      "datasource": {
        "type": "loki",
        "uid": "loki"
      },
      "gridPos": {
        "h": 12,
        "w": 24,
        "x": 0,
        "y": 12
      },
      "id": 4,
      "options": {
        "showTime": true,
        "showLabels": false,
        "showCommonLabels": false,
        "wrapLogMessage": false,
        "prettifyLogMessage": false,
        "enableLogDetails": true,
        "dedupStrategy": "none",
        "sortOrder": "Descending"
      },
      "pluginVersion": "10.3.3",
      "targets": [
        {
          "datasource": {
            "type": "loki",
            "uid": "loki"
          },
          "expr": "{job=~\"stepflow/.*\"}",
          "refId": "A"
        }
      ],
      "title": "Stepflow Logs",
      "type": "logs"
    }
  ],
  "schemaVersion": 39,
  "tags": [
    "stepflow",
    "workflows"
  ],
  "templating": {
    "list": []
  },
  "time": {
    "from": "now-6h",
    "to": "now"
  },
  "timepicker": {},
  "timezone": "",
  "title": "Stepflow Workflow Overview",
  "uid": "stepflow-workflow-overview",
  "version": 1,
  "weekStart": ""
}
