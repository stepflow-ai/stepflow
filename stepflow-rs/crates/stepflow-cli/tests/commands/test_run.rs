// Copyright 2025 DataStax Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
// in compliance with the License. You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software distributed under the License
// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
// or implied. See the License for the specific language governing permissions and limitations under
// the License.

use super::stepflow;
use insta_cmd::assert_cmd_snapshot;

#[test]
fn test_run_help() {
    assert_cmd_snapshot!(stepflow().arg("run").arg("--help"));
}

#[test]
fn test_run_basic_workflow_with_file_input() {
    assert_cmd_snapshot!(
        stepflow()
            .arg("run")
            .arg("--flow=tests/mock/basic.yaml")
            .arg("--input=tests/mock/input.json")
    );
}

#[test]
fn test_run_basic_workflow_with_json_input() {
    assert_cmd_snapshot!(
        stepflow()
            .arg("run")
            .arg("--flow=tests/mock/basic.yaml")
            .arg(r#"--input-json={"name": "hello"}"#)
    );
}

#[test]
fn test_run_basic_workflow_with_yaml_input() {
    assert_cmd_snapshot!(
        stepflow()
            .arg("run")
            .arg("--flow=tests/mock/basic.yaml")
            .arg("--input-yaml=name: hello")
    );
}

#[test]
fn test_run_with_custom_config() {
    assert_cmd_snapshot!(
        stepflow()
            .arg("run")
            .arg("--flow=tests/builtins/blob_test.yaml")
            .arg("--config=tests/builtins/stepflow-config.yml")
            .arg(r#"--input-json={"data": "hello"}"#)
    );
}

#[test]
fn test_run_nonexistent_workflow() {
    assert_cmd_snapshot!(
        stepflow()
            .arg("run")
            .arg("--flow=nonexistent.yaml")
            .arg(r#"--input-json={}"#)
    );
}

#[test]
fn test_run_invalid_input_json() {
    assert_cmd_snapshot!(
        stepflow()
            .arg("run")
            .arg("--flow=tests/mock/basic.yaml")
            .arg("--input-json=invalid json")
    );
}

#[test]
fn test_run_empty_flow() {
    assert_cmd_snapshot!(
        stepflow()
            .arg("run")
            .arg("--flow=tests/empty/empty_flow.yaml")
            .arg("--config=tests/empty/stepflow-config.yml")
            .arg(r#"--input-json={"message": "hello world", "numbers": {"a": 10, "b": 20}, "items": ["first", "second", "third"]}"#)
    );
}
