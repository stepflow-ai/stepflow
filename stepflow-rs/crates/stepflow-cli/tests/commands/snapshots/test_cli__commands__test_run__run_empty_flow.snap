---
source: crates/stepflow-cli/tests/commands/test_run.rs
info:
  program: stepflow
  args:
    - "--log-file=/dev/null"
    - "--omit-stack-trace"
    - run
    - "--flow=tests/empty/empty_flow.yaml"
    - "--config=tests/empty/stepflow-config.yml"
    - "--input-json={\"message\": \"hello world\", \"numbers\": {\"a\": 10, \"b\": 20}, \"items\": [\"first\", \"second\", \"third\"]}"
  env:
    STEPFLOW_LOG_DESTINATION: ""
    STEPFLOW_LOG_FILE: ""
    STEPFLOW_LOG_FORMAT: ""
    STEPFLOW_LOG_LEVEL: ""
    STEPFLOW_OTHER_LOG_LEVEL: ""
    STEPFLOW_OTLP_ENDPOINT: ""
    STEPFLOW_TRACE_ENABLED: ""
---
success: true
exit_code: 0
----- stdout -----
üìä Validation results: 0 fatal, 0 errors, 5 warnings

‚ö†Ô∏è  Warnings:
  WARN Field access '$.message' on step 'workflow_input' cannot be validated: no input schema available (output.message)
  WARN Field access '$.numbers.a' on step 'workflow_input' cannot be validated: no input schema available (output.a_value)
  WARN Field access '$.numbers.b' on step 'workflow_input' cannot be validated: no input schema available (output.b_value)
  WARN Field access '$.items[0]' on step 'workflow_input' cannot be validated: no input schema available (output.first_item)
  WARN Field access '$.numbers' on step 'workflow_input' cannot be validated: no input schema available (output.nested_data)
{"outcome":"success","result":{"message":"hello world","a_value":10,"b_value":20,"first_item":"first","nested_data":{"a":10,"b":20}}}
----- stderr -----
run_id: [ID]
